app/src/auth/hooks/useWalletConnect.js

```diff
@@ -0,0 +1,134 @@
+                },
+                signature,
+                address,
+            },
+        })
+    } catch (error) {
+        console.warn(error)
+        throw new ChallengeFailedError()
+    }
+}
+
+const useWalletConnect = () => {
+    const connect = useCallback(async () => (
+        new Promise((resolve, reject) => {
+            // Create a connector
+            let connector = new WalletConnect({
+                bridge: 'https://bridge.walletconnect.org', // Required
+                qrcodeModal: QRCodeModal,
+            })
+
+            let challengePending = false
+            let connected = false
+            let finalized = false
+
+            const finalize = async ({ token, error, disconnect = true } = {}) => {
+                finalized = true
+
+                if (disconnect && connected) {
+                    await connector.killSession()
+                    connector = undefined
+                    connected = false
+                }
+
+                if (token) {
+                    resolve(token)
+                } else {
+                    reject(error)
+                }
+            }
+
+            const onConnect = (error, payload) => {
+                if (!connected) {
+                    connected = true
+                }
+
+                if (error) {
+                    reject(error)
+                }
+
+                if (!challengePending) {
+                    // Get provided accounts and chainId
+                    const { accounts /* , chainId */ } = payload.params[0]
+
+                    challengePending = true
+
+                    signChallenge({
+                        address: accounts[0],
+                        connector,
+                    }).then((response) => (
+                        finalize({
+                            token: response && response.token,
+                        })
+                    ), (e) => (
+                        finalize({
+                            error: e,
+                        })
+                    ))
+                }
+            }
+
+            const onDisconnect = (error) => {
+                if (!finalized) {
+                    finalize({
+                        error: error || new Error('Session disconnected'),
+                        disconnect: false,
+                    })
+                }
+            }
+
+            // Check if connection is already established
+            if (!connector.connected) {
+                // create new session
+                connector.createSession()
+            } else {
+                onConnect(undefined, {
+                    params: [connector],
+                })
+            }
+
+            if (!connector) {
+                throw new Error('connector hasn\'t been created yet')
+            }
+
+            // Subscribe to connection events
+            connector.on('connect', onConnect)
+            connector.on('disconnect', onDisconnect)
```

Comment:
There is no apparent way to unsubscribe from these events 🤷

Reply:
Yep, tried `off`, `removeListener` and `removeAllListeners` but they don't exist. I believe the point of WalletConnect is to persist the session for the duration (it keeps polling the account). Here we are killing the session after getting the signature, I suppose it's not designed for one-off usage.